lib/min_heap.rb

```diff
@@ -44,29 +50,54 @@ def to_s
   end
 
   # This method returns true if the heap is empty
-  # Time complexity: ?
-  # Space complexity: ?
+  # Time complexity: o(1)
+  # Space complexity: o(1)
   def empty?
-    raise NotImplementedError, "Method not implemented yet..."
+    # raise NotImplementedError, "Method not implemented yet..."
+    return @store.empty?
   end
 
   private
 
   # This helper method takes an index and
   #  moves it up the heap, if it is less than it's parent node.
   #  It could be **very** helpful for the add method.
-  # Time complexity: ?
-  # Space complexity: ?
+  # Time complexity: o(logn)
+  # Space complexity: o(1)
   def heap_up(index)
-    
+    parent_index = (index - 1) / 2
+    if @store[parent_index].key > @store[index].key && index != 0
+      swap(parent_index, index)
+      heap_up(parent_index)
+    end
   end
 
   # This helper method takes an index and 
   #  moves it up the heap if it's smaller
   #  than it's parent node.
   def heap_down(index)
-    raise NotImplementedError, "Method not implemented yet..."
+    # raise NotImplementedError, "Method not implemented yet..."
+    left_index = (index * 2) + 1
+    right_index = (index * 2) + 2
+
+    return if @store[left_index].nil?
+
+    if @store[right_index].nil?
+      if @store[index].value > @store[left_index].value
+        swap(index, left_index)
+      end
+      return
+    end
+
+    if @store[left_index].key < @store[right_index].key
```

Comment:
You also need to compare the child to the current element. 

```suggestion
    if @store[left_index].key < @store[right_index].key && @store[left_index].key < @store[index].key
```